src/equal.c

```diff
@@ -418,6 +421,162 @@ bool equal_names(SEXP x, SEXP y) {
+}
+
+// -----------------------------------------------------------------------------
+
+#define EQUAL_COL(CTYPE, CONST_DEREF, SCALAR_EQUAL)                  \
+do {                                                                 \
+  int* p_out = LOGICAL(info.out);                                    \
+  int* p_row_known = LOGICAL(info.row_known);                        \
+                                                                     \
+  const CTYPE* p_x = CONST_DEREF(x);                                 \
+  const CTYPE* p_y = CONST_DEREF(y);                                 \
+                                                                     \
+  for (R_len_t i = 0; i < n_row; ++i, ++p_row_known, ++p_x, ++p_y) { \
+    if (*p_row_known) {                                              \
+      continue;                                                      \
+    }                                                                \
+                                                                     \
+    int eq = SCALAR_EQUAL(p_x, p_y, na_equal);                       \
+                                                                     \
+    if (eq <= 0) {                                                   \
+      p_out[i] = eq;                                                 \
+      *p_row_known = true;                                           \
+      --info.remaining;                                              \
+                                                                     \
+      if (info.remaining == 0) {                                     \
+        break;                                                       \
+      }                                                              \
+    }                                                                \
+  }                                                                  \
+                                                                     \
+  return info;                                                       \
+}                                                                    \
+while (0)
+
+#define EQUAL_COL_BARRIER(SCALAR_EQUAL)                \
+do {                                                   \
+  int* p_out = LOGICAL(info.out);                      \
+  int* p_row_known = LOGICAL(info.row_known);          \
+                                                       \
+  for (R_len_t i = 0; i < n_row; ++i, ++p_row_known) { \
+    if (*p_row_known) {                                \
+      continue;                                        \
+    }                                                  \
+                                                       \
+    int eq = SCALAR_EQUAL(x, i, y, i, na_equal);       \
+                                                       \
+    if (eq <= 0) {                                     \
+      p_out[i] = eq;                                   \
+      *p_row_known = true;                             \
```

Comment:
I think this should be `p_row_known[i]`, otherwise it suggests that it has a different increment rule than for `p_out` which is indexed with `i`.

Reply:
Would it make sense for `row_known` to just be a `bool` array that we `R_alloc()`? I don't use the fact that it is a SEXP anywhere, and it might make this clearer:

```
struct vctrs_df_rowwise_info {
  SEXP out;
  SEXP row_known;
  R_len_t remaining;
};
```

```
struct vctrs_df_rowwise_info {
  SEXP out;
  bool* row_known;
  R_len_t remaining;
};
```

It would also nicely enforce the type that `row_known` is intended to be. Right now it can be any SEXP. This is a nice property of `out`, but not of `row_known`.

Reply:
I think that makes sense, we could try this pattern. It is easy to set and reset the vmax from our `PROTECT_` function right?